fix panic in container_handler conversion func

cgtexmex opened the pull request:
Checks for existence of session id information in
execConfig prior to interrogating map

fixes #2985

hmahmood commented on lib/apiservers/portlayer/restapi/handlers/containers_handlers.go:
```diff
@@ -436,28 +428,41 @@ func convertContainerToContainerInfo(container *exec.ContainerInfo) *models.Cont
+	// session id in execConfig -- this has only manifest itself in short lived containers
+	// that were initilized via run
+	if _, exists := container.ExecConfig.Sessions[ccid]; exists {
+		tty := container.ExecConfig.Sessions[ccid].Tty
```
Don't need to do `container.ExecConfig.Sessions[ccid]` again here; it has already been done on the line above.

cgtexmex replied:
done

matthewavery commented on lib/apiservers/portlayer/restapi/handlers/containers_handlers.go:
```diff
@@ -436,28 +428,41 @@ func convertContainerToContainerInfo(container *exec.ContainerInfo) *models.Cont
 
-	path := container.ExecConfig.Sessions[ccid].Cmd.Path
-	info.ProcessConfig.ExecPath = &path
+	// in heavily loaded environments we were seeing a panic due a missing
```
`...due to...` sorry :) just saw it and thought now was a good time.

matthewavery commented on lib/apiservers/portlayer/restapi/handlers/containers_handlers.go:
```diff
@@ -436,28 +428,41 @@ func convertContainerToContainerInfo(container *exec.ContainerInfo) *models.Cont
-	path := container.ExecConfig.Sessions[ccid].Cmd.Path
-	info.ProcessConfig.ExecPath = &path
+	// in heavily loaded environments we were seeing a panic due a missing
+	// session id in execConfig -- this has only manifest itself in short lived containers
```
`manifested`*

matthewavery commented:
lgtm

caglar10ur commented on lib/apiservers/portlayer/restapi/handlers/containers_handlers.go:
```diff
@@ -436,28 +428,41 @@ func convertContainerToContainerInfo(container *exec.ContainerInfo) *models.Cont
-	info.ProcessConfig.ExecPath = &path
+	// in heavily loaded environments we were seeing a panic due to a missing
+	// session id in execConfig -- this has only manifested itself in short lived containers
+	// that were initilized via run
```
I think I didn't understand this comment. Do we know why session id is missing? It sounds like a corruption or race somewhere and if that's the case this is just a workaround

cgtexmex replied:
No - I don't know the exact cause of the issue -- to your point I have only avoided the panic.  This happens in the CI intermittently and when the environment is apparently loaded / slow -- the test failed due to the panic, but would have failed because of network issues.

We can leave this change out and leave the issue open -- but reproduction is a challenge.  I was banking on the avoidance of a panic being the preferred approach.

hickeng replied:
We can compromise by protecting against the panic if DebugLevel == 0.

Irrespective, we should log the actual contents - it's possible the entire extraconfig wasn't populated/unmarshalled correctly.

cgtexmex commented:
Updated with logging of the ExecConfig is sessionID is missing.  Additionally, introduced a minor refactoring to the eventCallback to avoid duplicating container refresh during the container wait path.

vburenin commented:
LGTM

hickeng commented:
LGTM - please open an issue to track down why we're getting nil session content.

mdubya66 commented:
lgtm recycle
